File: src/main/scala/stni/atlassian/remote/query/QueryIssueList.scala
package stni.atlassian.remote.query

import com.atlassian.jira.rpc.soap.beans.{RemoteIssueType, RemotePriority, RemoteIssue}
import JqlBuilder._
import collection.convert.Wrappers

/**
 *
 */
class QueryIssueList(query: JiraQuery, elems: Seq[QueryIssue]) extends QueryIssueWrapper(elems, qi => new QueryIssueList(query, List(qi))) {

  import QueryIssueList._

  def children(expression: String, order: String): QueryIssueList =
    if (this.isEmpty) {
      ofQueryIssues(query)
    } else {
      ofJql(query, parentIn(this), expression, order)
    }

  def typedLinked(linkType: String, expression: String, order: String): QueryIssueList = {
    ofRemoteIssues(query, this.flatMap(elem => query.executeJql(jql(linkedWith(elem.getKey, linkType), expression, order))))
  }

  def asJList: java.util.List[QueryIssueList] = Wrappers.SeqWrapper(this)

  def formattedCustomField(name: String): String = query.parseAndFormat(customField(name), this)

  def customField(name: String): String = query.getCustomField(issue, name)

  def summary: String = query.formatAsPlain(super.getSummary)

  def description: String = query.parseAndFormat(super.getDescription, this)

  def url: String = url(getKey)

  def url(otherIssueKey: String): String = query.getBaseUrl + "/browse/" + otherIssueKey

  def priority: RemotePriority = query.priorityById(getPriority)

  def issueType: RemoteIssueType = query.issueTypeById(getType)

  def timeToResolve: Long = query.timeToResolve(issue)

}

object QueryIssueList {
  def ofQueryIssues(query: JiraQuery, elems: QueryIssue*): QueryIssueList = new QueryIssueList(query, elems)

  def ofJql(query: JiraQuery, context: String, expression: String, order: String): QueryIssueList =
    ofRemoteIssues(query, query.executeJql(jql(context, expression, order)))

  def ofRemoteIssues(query: JiraQuery, issues: Seq[RemoteIssue]): QueryIssueList =
    ofQueryIssues(query, issues.map(ri => new QueryIssue(ri)): _*)

}

File: src/main/java/stni/atlassian/remote/queryold/QueryIssueList.java
package stni.atlassian.remote.queryold;

import com.atlassian.jira.rpc.soap.beans.RemoteIssue;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static stni.atlassian.remote.queryold.JqlBuilder.*;

/**
 *
 */
final class QueryIssueList extends AbstractQueryList<QueryIssue> {
    QueryIssueList(JiraQuery query, List<QueryIssue> issues) {
        super(query,issues);
    }

    public QueryList<QueryIssue> newQueryList(JiraQuery query, List<QueryIssue> issues) {
        return new QueryIssueList(query, issues);
    }

    static QueryList<QueryIssue> ofRemoteIssues(JiraQuery query, RemoteIssue... issues) {
        return ofRemoteIssueList(query, Arrays.asList(issues));
    }

    static QueryList<QueryIssue> ofRemoteIssueList(JiraQuery query, List<RemoteIssue> issues) {
        List<QueryIssue> list = new ArrayList<QueryIssue>(issues.size());
        for (RemoteIssue issue : issues) {
            list.add(new QueryIssue(query, issue));
        }
        return new QueryIssueList(query, list);
    }

    static QueryList<QueryIssue> ofJql(JiraQuery query, String context, String expression, String order) {
        return ofRemoteIssues(query, query.executeJql(jql(context, expression, order)));
    }

    public QueryList<QueryIssue> children(String expression, String order) {
        if (getElements().isEmpty()) {
            return newQueryList(getQuery(), null);
        }
        return ofJql(getQuery(), parentIn(getElements()), expression, order);
    }


    public QueryList<QueryIssue> typedLinked(String linkType, String expression, String order) {
        List<RemoteIssue> res = new ArrayList<RemoteIssue>();
        for (QueryIssue issue : getElements()) {
            res.addAll(Arrays.asList(getQuery().executeJql(jql(linkedWith(issue.getKey(), linkType), expression, order))));
        }
        return ofRemoteIssueList(getQuery(), res);
    }
}
